apps/backend/src/app/modules/users/users.service.spec.ts

```diff
@@ -0,0 +1,286 @@
+        avatarhash: '',
+        avatarmedium: '',
+        communityvisibilitystate: 0,
+        lastlogoff: 0,
+        loccountrycode: '',
+        personaname: '',
+        personastate: 0,
+        personastateflags: 0,
+        primaryclanid: '',
+        profilestate: 1,
+        profileurl: '',
+        realname: '',
+        steamid: '123456789',
+        timecreated: 0
+      });
+      expect(spy).toHaveBeenCalled();
+    });
+
+    it('should error when no user is returned by findOrCreateUser', async () => {
+      const spy = jest.spyOn(usersService, 'findOrCreateUser');
+      spy.mockImplementationOnce(() => Promise.resolve(undefined));
+      await expect(
+        usersService.findOrCreateFromWeb({
+          avatar: '',
+          avatarfull: '',
+          avatarhash: '',
+          avatarmedium: '',
+          communityvisibilitystate: 0,
+          lastlogoff: 0,
+          loccountrycode: '',
+          personaname: '',
+          personastate: 0,
+          personastateflags: 0,
+          primaryclanid: '',
+          profilestate: 1,
+          profileurl: '',
+          realname: '',
+          steamid: '123456789',
+          timecreated: 0
+        })
+      ).rejects.toThrow(InternalServerErrorException);
+    });
+  });
+
+  describe('findOrCreateUser', () => {
+    it('should return an existing user', async () => {
+      const existingUser: User = {
+        id: 0,
+        roles: 0,
+        bans: 0,
+        steamID: BigInt(123456789),
+        alias: '',
+        avatar: '',
+        country: '',
+        createdAt: undefined
+      };
+      db.user.findUnique.mockResolvedValueOnce(existingUser as any);
+      const authUser: AuthenticatedUser = {
+        id: 0,
+        steamID: 123456789n
+      };
+      db.user.update.mockResolvedValueOnce(authUser as any);
+
+      const userData = {
+        steamID: BigInt(123456789),
+        alias: 'new alias',
+        avatar: 'new avatar',
+        country: 'new country'
+      };
+      const spy = jest.spyOn(db['user'], 'update');
+      await usersService.findOrCreateUser(userData);
+      expect(spy).toHaveBeenCalled();
+    });
+    it('should create a new user', async () => {
+      const nonExistingUser: User = undefined;
+      db.user.findUnique.mockResolvedValueOnce(nonExistingUser as any);
+
+      const spy = jest.spyOn(db['user'], 'create');
+
+      const userData = {
+        steamID: BigInt(123456789),
+        alias: '',
+        avatar: '',
+        country: ''
+      };
+      await usersService.findOrCreateUser(userData);
+
+      expect(spy).toHaveBeenCalled();
+    });
+    it('should throw error when using a limited account if the setting is on', async () => {
+      const nonExistingUser: User = undefined;
+      db.user.findUnique.mockResolvedValueOnce(nonExistingUser as any);
+
+      jest
+        .spyOn(usersService['config'], 'getOrThrow')
+        .mockImplementationOnce(() => {
+          return true;
+        });
```

Comment:
I wouldn't usually bring this up but since I think you said you were new to JS, I just want to point out that this is equivalent to `.mockImplementationOnce(() => true )` - if an arrow function is a single expression (no block/semicolons) it just returns the result of the expression,

Reply:
Oh actually, this can just be `mockReturnValueOnce(true)`, as you're doing in some places above!

Reply:
Ah, thanks for pointing that out! `mockReturnValueOnce` is definitely nicer in these situations.